### PR TITLE
🐛 Fixed removing comped subscriptions for members with active subs

### PR DESCRIPTION
**Pull request by SimonBackx**
fixes https://github.com/TryGhost/Team/issues/1859

**Problem:**
When for some reason a member has an active subscription (or legacy comped subscription) for product A, and a comped subscription for product B. You cannot remove comped subscription B.

**Fixed by:**
Updating the API to allow more flexible product changes on members.
- Allow the removal of (comped) products on a member, as long as that product doesn't have a related subscription
- (still) allow the addition of comped products to a member, as long as that member doesn't have other active subscriptions. This matches the existing behaviour, but now this is only checked for added products.
- Includes tests for these edge cases

### CODE REVIEW

File: ghost/core/test/e2e-api/admin/members.test.js [hunk 1]
```diff
@@ -1384,6 +1388,126 @@ describe('Members API', function () {
         should.deepEqual(memberWithPaidSubscription, readMember, 'Editing a member returns a different format than reading a member');
     });
 
+    it('Cannot add complimentary subscriptions to a member with an active subscription', async function () {
+        if (!memberWithPaidSubscription) {
+            // Previous test failed
+            this.skip();
```
Comment (allouis) [inline ghost/core/test/e2e-api/admin/members.test.js:1394]: This is an interesting pattern, have you seen this used elsewhere in the tests?

I'm not against it, as it is frustrating when a single test failing causes a cascade of later ones failing.

Ultimately that's because we have tests depending on some global state outside of the test, which IMO is not good.

We _could_ fix this by making sure each test does the setup that it needs, but that adds overhead.

Using this pattern at least makes the coupling explicit 🤔 Though what would happen if we removed the previous test, it would render this test useless right?

I'd be interested in your thoughts either way on this, and if you think it's a good idea - let's talk with Hannah about sharing/using this pattern going forwards

File: ghost/core/test/e2e-api/admin/members.test.js [hunk 2]
```diff
@@ -1384,6 +1388,126 @@ describe('Members API', function () {
...
+            .expectStatus(200);
+
+        const beforeMember = body2.members[0];
+        assert.equal(beforeMember.tiers.length, 2, 'The member should have two products now');
+
+        // Now try to remove only the complimentary one
+        const compedPayload = {
+            id: memberWithPaidSubscription.id,
+            // Remove all complimentary subscriptions
+            tiers: [...memberWithPaidSubscription.tiers]
```
Comment (allouis) [inline ghost/core/test/e2e-api/admin/members.test.js:1468]: No need for spread operator here, right? NABD

File: ghost/core/test/e2e-api/admin/members.test.js [hunk 3]
```diff
@@ -1384,6 +1388,126 @@ describe('Members API', function () {
...
+            .put(`/members/${memberWithPaidSubscription.id}/`)
+            .body({members: [compedPayload]})
+            .expectStatus(200);
+
+        const updatedMember = body.members[0];
+        assert.equal(updatedMember.status, 'paid', 'Member should still have the paid status');
+        assert.equal(updatedMember.tiers.length, 1, 'The member should have one product now');
+        assert.equal(updatedMember.tiers[0].id, memberWithPaidSubscription.tiers[0].id, 'The member should have the paid product');
+
+        /*const compedPayload = {
```
Comment (allouis) [inline ghost/core/test/e2e-api/admin/members.test.js:1481]: This is the same as the previous test right? Let's remove the comment, or uncomment and remove previous test - up to you ☺️

**Review by allouis — APPROVED** [review]
Couple comments but otherwise 👍

File: ghost/core/test/e2e-api/admin/members.test.js (hunk 1 above)
Comment (SimonBackx) [reply ghost/core/test/e2e-api/admin/members.test.js:1394]: Yeah, I introduced this pattern a couple of months ago in the same file. I didn't really think a lot about it when writing these additional tests. I don't think this pattern is ideal, but it was an effort to prevent having to repeat all the Stripe mocking in every test. But moving that repeated part to the top might be a better solution... 🤔 Or we could build the models manually without mocking Stripe...

File: ghost/core/test/e2e-api/admin/members.test.js (hunk 2 above)
Comment (SimonBackx) [reply ghost/core/test/e2e-api/admin/members.test.js:1468]: Right 😅👌
```suggestion
            tiers: memberWithPaidSubscription.tiers
```

File: ghost/core/test/e2e-api/admin/members.test.js (hunk 3 above)
Comment (SimonBackx) [reply ghost/core/test/e2e-api/admin/members.test.js:1481]: Copy paste issue or something 😬

File: ghost/core/test/e2e-api/admin/members.test.js [hunk 4]
```diff
@@ -1384,6 +1388,126 @@ describe('Members API', function () {
...
+            id: memberWithPaidSubscription.id,
+            // Remove all paid subscriptions (= not allowed atm)
+            tiers: []
+        };
+
+        await agent
+            .put(`/members/${memberWithPaidSubscription.id}/`)
+            .body({members: [compedPayload]})
+            .expectStatus(400);*/
```
Comment (SimonBackx) [inline ghost/core/test/e2e-api/admin/members.test.js:1490]:
```suggestion
```

File: ghost/core/test/e2e-api/admin/members.test.js (hunk 1 above)
Comment (SimonBackx) [reply ghost/core/test/e2e-api/admin/members.test.js:1394]: Gonna move this to a tech debt issue and clean up the whole test file in one go 🧹

Comment (naz) [reply ghost/core/test/e2e-api/admin/members.test.js:1394]: @SimonBackx what was the outcome of the cleanup? If the `memberWithPaidSubscription` is used in a group of test, like it seems to be done here, probably best would be grouping such tests into a "describe" and having a `memberWithPaidSubscription` be set up through fixtures? 🤔
